laghos_solver.cpp

```diff
@@ -563,39 +563,113 @@ void LagrangianHydroOperator::ComputeDensity(ParGridFunction &rho) const
+                  vmag += pow(f_vals(k,q,e),ln_norm);
+               }
+               I(q,e) = vmag;
+         }
       }
-      MPI_Comm comm = H1.GetParMesh()->GetComm();
-      MPI_Allreduce(&loc_ie, &glob_ie, 1, MPI_DOUBLE, MPI_SUM, comm);
    }
+   else if (DIM == 2){
+      MFEM_FORALL_2D(e, NE, Q1D, Q1D, 1,
+      {
+         MFEM_FOREACH_THREAD(qy,y,Q1D)
+         {
+            MFEM_FOREACH_THREAD(qx,x,Q1D)
+            {
+               const int q = qx + qy * Q1D;
+               double vmag = 0; 
+               for(int k = 0; k < VDIM; k++){
+                  vmag += pow(f_vals(k,q,e),ln_norm);
+               }
+               I(q,e) = vmag;
+            }
+         }
+      });
+   }
+   else if (DIM == 3){
+    MFEM_FORALL_3D(e, NE, Q1D, Q1D, Q1D,
+      {
+         MFEM_FOREACH_THREAD(qz,z,Q1D)
+         {
+            MFEM_FOREACH_THREAD(qy,y,Q1D)
+            {
+               MFEM_FOREACH_THREAD(qx,x,Q1D)
+               {
+                  const int q = qx + (qy + qz * Q1D) * Q1D;
+                  double vmag = 0; 
+               for(int k = 0; k < VDIM; k++){
+                  vmag += pow(f_vals(k,q,e),ln_norm);
+               }
+               I(q,e) = vmag;
+               }
+            }
+         }
+      });
+   
+}
+   const double integral = integrand * mass;
+return integral; 
+
+}
+double LagrangianHydroOperator::InternalEnergy(const ParGridFunction &gf) const
+{
+   double glob_ie = 0.0;
+   
+    // get the restriction and interpolator objects 
+    const QuadratureInterpolator* l2_interpolator = L2.GetQuadratureInterpolator(ir);
+    l2_interpolator->SetOutputLayout(QVectorLayout::byVDIM);
+    auto L2r = L2.GetElementRestriction(ElementDofOrdering::LEXICOGRAPHIC);
+    const int NQ = ir.GetNPoints(); 
+   Vector e_vector(NE*NQ), eintQ(NE*NQ);
+
+    // Get internal energy at the quadrature points
+    L2r->Mult(gf, e_vector);
+    l2_interpolator->Values(e_vector, eintQ);
+
+   double internal_energy = ComputeVolumeIntegral(dim,NE,NQ,Q1D,1,1.0,qdata.rho0DetJ0w,eintQ); 
+    
+   MPI_Allreduce(&internal_energy, &glob_ie, 1, MPI_DOUBLE, MPI_SUM, L2.GetParMesh()->GetComm());
+
    return glob_ie;
 }
 
 double LagrangianHydroOperator::KineticEnergy(const ParGridFunction &v) const
 {
-   double glob_ke = 0.0;
-   // This should be turned into a kernel so that it could be displayed in pa
-   if (!p_assembly)
-   {
-      double loc_ke = 0.5 * Mv_spmat_copy.InnerProduct(v, v);
-      MPI_Allreduce(&loc_ke, &glob_ke, 1, MPI_DOUBLE, MPI_SUM,
-                    H1.GetParMesh()->GetComm());
-   }
-   return glob_ke;
+  double glob_ke = 0.0;
+   
+    // get the restriction and interpolator objects 
+    const QuadratureInterpolator* h1_interpolator = H1.GetQuadratureInterpolator(ir);
+    h1_interpolator->SetOutputLayout(QVectorLayout::byVDIM);
+    auto H1r = H1.GetElementRestriction(ElementDofOrdering::LEXICOGRAPHIC);
+    const int NQ = ir.GetNPoints(); 
+   Vector e_vector(dim*NE*NQ), ekinQ(dim*NE*NQ);
```

Comment:
Shouldn't the E-vector be over dofs, not quad points? I'm guessing that in the next two calls the `Mult` writes at the correct indices, and the `Values` reads the same indices.

Reply:
You're saying that I should have written `e_vector(dim*NE*NQ)` as `e_vector(dim*NE)`? And that the next few lines will read/write from the first dim*NE block, so it didn't trigger a segfault or anything like that?

Reply:
I think it should be
```
const int ND = H1.GetFE(0)->GetDof();
Vector e_vector(dim*NE*ND);
```
And similar change in the internal energy function.